python/mspasspy/global_history/manager.py

```diff
@@ -0,0 +1,261 @@
+import os
+import yaml
+import pymongo
+import pyspark
+import dask.bag as daskbag
+
+from bson.objectid import ObjectId
+from mspasspy.ccore.utility import MsPASSError
+from datetime import datetime
+from dill.source import getsource
+
+def mspass_spark_map(self, func, *args, global_history=None, alg_name=None, parameters=None, **kwargs):
+    """
+     This decorator method add more functionaliy on the standard spark map method and be a part of member functions
+     in the spark RDD library. Instead of performing the normal map function, if user provides global history manager,
+     alg_name and parameters as input, the global history manager will log down the usage of the algorithm.
+
+    :param func: target function
+    :param global_history: a user specified global history manager
+    :type global_history: :class:`GlobalHistoryManager`
+    :param alg_name: a user specified alg_name for the map operation
+    :type alg_name: :class:`str`
+    :param parameters: a user specified parameters for the map operation
+    :type parameters: :class:`str`
+    :return: a spark `RDD` format of objects.
+    """
+    if global_history:
+        # get the whole lambda function
+        alg_string = getsource(func)
+        # truncate to get the algorithm name and parameters
+        alg = alg_string[alg_string.index(':')+1 : alg_string.index(')')]
+        
+        if not alg_name:
+            # get the algorithm name
+            alg_name = alg[:alg.index('(')].strip()
+        if not parameters:
+            # get the parameters
+            parameters = alg[alg.index('(')+1:].strip()
+        
+        # get the alg_id if exists, else create a new one
+        alg_id = global_history.get_alg_id(alg_name, parameters)
+        if not alg_id:
+            alg_id = ObjectId()
```

Comment:
It seems the `get_alg_id` related code can be put into the `logging` method. The `logging` method can then be defined as `logging(alg_name, parameters)`

Reply:
After revising the design doc, I found that `alg_id` is actually needed here.